### PR TITLE
feat: add support for push notification using sns

### PR DESCRIPTION
**Pull request by Harish-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed preliminary testing using the [test suite](../../apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected.
- [ ] I have updated the required [api docs](../../apps/api/docs/) as applicable.
- [ ] I have added/updated test cases to the [test suite](../../apps/api/OsmoX.postman_collection.json) as applicable

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [ ] Screenshots have been added (optional)
- [x] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [ ] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [ ] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**

This Pr add a new Provider for push notification - sns

**Related changes:**

- Add new provider
- Add new migration
- Add aws-sdk package in package.json

**Screenshots:**

```json
{
    // Assuming providerId 10 also has channelType 4
    "providerId": 10,
    "data": {
        "message": {
             "APNS_SANDBOX": "{\"aps\":{\"alert\":{\"title\":\"Hello World\",\"body\":\"This is a test message\"},\"sound\":\"default\"}}"
        },
        // "target": "arn:aws:sns:us-west-2:505884080245:endpoint/GCM/OQSHA-Android/7fb080a5-a5e9-38f0-98fe-ebd0396e0076" 
        "target": "arn:aws:sns:us-west-2:505884080245:endpoint/APNS_SANDBOX/OQSHA-iOS-Dev/57711edf-6265-32fc-be15-1a0cfdcad72e"
    }
}

```
image

```json
{
    // Assuming providerId 10 also has channelType 4
    "providerId": 10,
    "data": {
        "message": {
            "GCM": "{\"notification\":{\"title\":\"Test Notification\",\"body\":\"This is a test notification\"}}"},
        "target": "arn:aws:sns:us-west-2:505884080245:endpoint/GCM/OQSHA-Android/7fb080a5-a5e9-38f0-98fe-ebd0396e0076" 
        // "target": "arn:aws:sns:us-west-2:505884080245:endpoint/APNS_SANDBOX/OQSHA-iOS-Dev/57711edf-6265-32fc-be15-1a0cfdcad72e"
    }
}
```

image

**Query request and response:**

sample request
```json
{
    "providerId": 10,
    "data": {
        "title": "Test Notification",
        "body": "This is a Test Notification",
        "target": "arn:aws:sns:us-east-1:131638606831:endpoint/GCM/Tweelin_android_notifications/cc78381d-bbda-3d92-8fe1-54dc8bd2377c"
    }
}
```
sample response 
```json
{
    "status": "success",
    "data": {
        "notification": {
            "providerId": 10,
            "data": {
                "title": "Test Notification",
                "body": "This is a Test Notification",
                "target": "arn:aws:sns:us-east-1:131638606831:endpoint/GCM/Tweelin_android_notifications/cc78381d-bbda-3d92-8fe1-54dc8bd2377c"
            },
            "channelType": 9,
            "applicationId": 1,
            "createdBy": "sampleOsmoXApp",
            "updatedBy": "sampleOsmoXApp",
            "result": null,
            "id": 29,
            "deliveryStatus": 1,
            "createdOn": "2024-07-05T07:11:52.000Z",
            "updatedOn": "2024-07-05T07:11:52.000Z",
            "status": 1,
            "retryCount": 0
        }
    }
}
```

**Documentation changes:**

NA

**Test suite output:**

NA

**Pending actions:**

- Testing is left as for now oqsha push notification is not working 

**Additional notes:**

NA

## Summary by CodeRabbit

- **New Features**
  - Introduced a new push notification service using AWS SNS. Users can now receive push notifications with specified data (title, body, message).

### CODE REVIEW

File: apps/api/src/common/decorators/is-data-valid.decorator.ts [hunk 1]
```diff
@@ -95,6 +96,12 @@ export class IsDataValidConstraint implements ValidatorConstraintInterface {
         await validateAndThrowError(kapsystemData);
         return true;
 
+      case ChannelType.PUSH_SNS:
+        const pushSnsData = new PushSnsDataDto();
+        Object.assign(pushSnsData, value);
+        await validateAndThrowError(pushSnsData);
+        return true;
+
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/common/decorators/is-data-valid.decorator.ts:104]: If fine like this also to maintain consistency used like that

File: apps/api/src/modules/providers/push-sns/push-sns.service.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      accessKeyId: snsConfig.AWS_ACCESS_KEY_ID as string,
+      secretAccessKey: snsConfig.AWS_SECRET_ACCESS_KEY as string,
+      region: snsConfig.AWS_REGION as string,
+    });
+  }
+
+  async sendPushNotification(data: PushSnsData, providerId: number): Promise<SNS.PublishResponse> {
+    await this.assignSnsConfig(providerId);
+    const { title, body } = data;
+    const message = `${title}: ${body}`;
```
Comment (xixas) [inline apps/api/src/modules/providers/push-sns/push-sns.service.ts:30]: DId you see how you are getting notification with FCM when we use title and body?
Here from what I see we have kept title and body as necessary but then you are fixing the format of notification limiting the usecase of notification message.

Once check what effect title and body does from firebase. Based on your findings, we can just keep message as the field and remove title and body if they don't effect.

Here I see message as an object, you can test with this in SNS too and see its formatted correctly.
image

File: apps/api/package.json [hunk 3]
```diff
@@ -42,12 +42,14 @@
...
     "@nestjs/typeorm": "^10.0.2",
     "@types/bcrypt": "^5.0.2",
+    "aws-sdk": "^2.1654.0",
     "axios": "^1.7.2",
     "bcrypt": "^5.1.1",
     "bull": "^4.12.9",
     "bullmq": "^5.8.3",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.1",
+    "firebase-admin": "^12.2.0",
```
Comment (xixas) [inline apps/api/package.json:52]: Do we need this package for AWS SNS?

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/api/package.json (hunk 3 above)
Comment (Harish-osmosys) [reply apps/api/package.json:52]: This was added to check fcm will remove it

File: apps/api/src/modules/providers/push-sns/push-sns.service.ts (hunk 2 above)
Comment (Harish-osmosys) [reply apps/api/src/modules/providers/push-sns/push-sns.service.ts]: updated in https://github.com/OsmosysSoftware/osmo-x/pull/269/commits/c89ab05d02d57df395c3776d0070d1823d4712cd

File: apps/api/package.json (hunk 3 above)
Comment (Harish-osmosys) [reply apps/api/package.json]: removed in https://github.com/OsmosysSoftware/osmo-x/pull/269/commits/201e68ef2916657f9528af193ce588e661831389

File: apps/api/src/common/decorators/is-data-valid.decorator.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/common/decorators/is-data-valid.decorator.ts:104]: @Harish-osmosys  For now add a todo and ask Lakshya to create task, we can work on it later when needed.

**Review by xixas — APPROVED** [review]

**Pull request merged**